File: src/com/kingmang/lazurite/parser/AST/Statements/IfStatement.java
package com.kingmang.lazurite.parser.AST.Statements;


import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;

public final class IfStatement extends InterruptableNode implements Statement {
    
    public final Expression expression;
    public final Statement ifStatement, elseStatement;

    public IfStatement(Expression expression, Statement ifStatement, Statement elseStatement) {
        this.expression = expression;
        this.ifStatement = ifStatement;
        this.elseStatement = elseStatement;
    }
    
    @Override
    public void execute() {
        super.interruptionCheck();
        final int result = expression.eval().asInt();
        if (result != 0) {
            ifStatement.execute();
        } else if (elseStatement != null) {
            elseStatement.execute();
        }
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        final StringBuilder result = new StringBuilder();
        result.append("if ").append(expression).append(' ').append(ifStatement);
        if (elseStatement != null) {
            result.append("\nelse ").append(elseStatement);
        }
        return result.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/IfStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class IfStatement(
    val expression: Expression,
    val ifStatement: Statement,
    val elseStatement: Statement?
) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        val result = this.expression.eval().asInt()
        if (result != 0) {
            this.ifStatement.execute()
        } else this.elseStatement?.execute()
    }

    override fun accept(visitor: Visitor) {
        visitor.visit(this)
    }

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R? {
        return visitor.visit(this, t)
    }

    override fun toString(): String =
        StringBuilder().run {
            this.append("if ").append(expression).append(' ').append(ifStatement)
            elseStatement?.let { this.append("\nelse ").append(it) }
            return this.toString()
        }
}
